src/jsTypeFactory.cc

```diff
@@ -140,13 +155,18 @@ JS::Value jsTypeFactory(JSContext *cx, PyObject *object) {
     }
 
     JSFunction *jsFunc = js::NewFunctionWithReserved(cx, callPyFunc, nargs, 0, NULL);
-    JSObject *jsFuncObject = JS_GetFunctionObject(jsFunc);
-
+    JS::RootedObject jsFuncObject(cx, JS_GetFunctionObject(jsFunc));
     // We put the address of the PyObject in the JSFunction's 0th private slot so we can access it later
     js::SetFunctionNativeReserved(jsFuncObject, 0, JS::PrivateValue((void *)object));
     returnType.setObject(*jsFuncObject);
-    memoizePyTypeAndGCThing(new FuncType(object), returnType);
     Py_INCREF(object); // otherwise the python function object would be double-freed on GC in Python 3.11+
+
+    // add function to jsFunctionRegistry, to DECREF the PyObject when the JSFunction is finalized
+    JS::RootedValueArray<2> registerArgs(GLOBAL_CX);
+    registerArgs[0].setObject(*jsFuncObject);
+    registerArgs[1].setPrivate(object);
+    JS::RootedValue ignoredOutVal(GLOBAL_CX);
+    JS_CallFunctionName(GLOBAL_CX, *jsFunctionRegistry, "register", registerArgs, &ignoredOutVal);
```

Comment:
Can we check the return value of the JS_CallFunctionName call?

Reply:
done